haskell/src/Lentille/GraphQL.hs

```diff
@@ -110,17 +136,55 @@ data RateLimit = RateLimit {used :: Int, remaining :: Int, resetAt :: UTCTime}
 instance From RateLimit Text where
   from RateLimit {..} = "remains:" <> show remaining <> ", reset at: " <> show resetAt
 
+type RetryCheck m a = (Either (FetchError a) a, [ReqLog]) -> m Bool
+
+-- | wrapper around fetchWithLog than can optionaly handle fetch retries
+-- based on the returned data inspection via a provided function (see RetryCheck).
+-- In case of retry the depth parameter of mkArgs is decreased (see adaptDepth)
+doRequest ::
+  forall a.
+  forall m.
+  (MonadGraphQLE m, Fetch a, FromJSON a) =>
+  GraphClient ->
+  (Maybe Int -> Maybe Text -> Args a) ->
+  Maybe (RetryCheck m a) ->
+  Maybe Int ->
+  Maybe PageInfo ->
+  m (Either (FetchError a) a, [ReqLog])
+doRequest client mkArgs retryCheckM depthM pageInfoM = gRetry retryCheck runFetch
+  where
+    gRetry = genericRetry Macroscope "Retrying request with smaller depth"
+    retryCheck _ = fromMaybe (const $ pure False) retryCheckM
+    runFetch :: (MonadGraphQLE m) => Int -> m (Either (FetchError a) a, [ReqLog])
+    runFetch retried = do
+      let adapedDepth = adaptDepth
+      if adapedDepth == Just 0
+        then error "Unable to reduce depth more"
+        else
+          fetchWithLog
+            (doGraphRequest client)
+            (mkArgs depthM $ (Just . fromMaybe (error "Missing endCursor from page info") . endCursor) =<< pageInfoM)
+      where
+        adaptDepth :: Maybe Int
+        adaptDepth =
+          let decValue = truncate @Float . (* (fromIntegral retried * 0.3)) . fromIntegral <$> depthM
+           in (-) <$> depthM <*> decValue
```

Comment:
might be good to factor the non monadic part as a top-level function with some doctest such as:

```suggestion
        adaptDepth :: Maybe Int
        adaptDepth = decreaseValue retried <$> depthM
        
-- | Slowly decrease a value to workaround api timeout when a graph depth is too deep.
 -- >>> decreaseValue 1 42
 -- 29
 -- >>> decreaseValue 1 29
 -- 20
decreaseValue :: Int -> Int -> Int
decreaseValue retried depth =
          let decValue = truncate @Float . (* (fromIntegral retried * 0.3)) . fromIntegral $ depth
           in depth - decValue
```

Reply:
Shouldn't the decreaseValue also clamp at 0?, e.g. `max 0 $ depth - decValue`